// zmp-ui/cjs/components/button/props-type.js
"use strict";

exports.__esModule = true;
exports.ButtonVariants = exports.ButtonType = exports.ButtonSize = exports.ButtonHTMLType = void 0;

/**
 * @category General
 * @subcategory Button
 * @enum {string}
 */
var ButtonType;
/**
 * @category General
 * @subcategory Button
 * @enum {string}
 */

exports.ButtonType = ButtonType;

(function (ButtonType) {
  ButtonType["highlight"] = "highlight";
  ButtonType["danger"] = "danger";
  ButtonType["neutral"] = "neutral";
})(ButtonType || (exports.ButtonType = ButtonType = {}));

var ButtonHTMLType;
/**
 * @category General
 * @subcategory Button
 * @enum {string}
 */

exports.ButtonHTMLType = ButtonHTMLType;

(function (ButtonHTMLType) {
  ButtonHTMLType["submit"] = "submit";
  ButtonHTMLType["button"] = "button";
  ButtonHTMLType["reset"] = "reset";
})(ButtonHTMLType || (exports.ButtonHTMLType = ButtonHTMLType = {}));

var ButtonSize;
/**
 * @category General
 * @subcategory Button
 * @enum {string}
 */

exports.ButtonSize = ButtonSize;

(function (ButtonSize) {
  ButtonSize["large"] = "large";
  ButtonSize["medium"] = "medium";
  ButtonSize["small"] = "small";
})(ButtonSize || (exports.ButtonSize = ButtonSize = {}));

var ButtonVariants;
exports.ButtonVariants = ButtonVariants;

(function (ButtonVariants) {
  ButtonVariants["primary"] = "primary";
  ButtonVariants["secondary"] = "secondary";
  ButtonVariants["tertiary"] = "tertiary";
})(ButtonVariants || (exports.ButtonVariants = ButtonVariants = {}));
