// zmp-ui/esm/components/snackbar-provider/props-type.js
/* eslint-disable no-use-before-define */

/**
 * @category Feedback
 * @subcategory Snackbar
 * @enum {string}
 */
export var SnackbarPosition;
/**
 * @category Feedback
 * @subcategory Snackbar
 * @enum {string}
 */

(function (SnackbarPosition) {
  SnackbarPosition["top"] = "top";
  SnackbarPosition["bottom"] = "bottom";
})(SnackbarPosition || (SnackbarPosition = {}));

export var SnackbarType;
/**
 * @category Feedback
 * @subcategory Snackbar
 * @displayName SnackbarProvider
 */

(function (SnackbarType) {
  SnackbarType["default"] = "default";
  SnackbarType["success"] = "success";
  SnackbarType["info"] = "info";
  SnackbarType["error"] = "error";
  SnackbarType["warning"] = "warning";
  SnackbarType["loading"] = "loading";
  SnackbarType["download"] = "download";
  SnackbarType["countdown"] = "countdown";
  SnackbarType["wifi-connected"] = "wifi-connected";
  SnackbarType["wifi-disconnected"] = "wifi-disconnected";
})(SnackbarType || (SnackbarType = {})); // NOT INJECT TYPE DECLARE
