// zmp-ui/cjs/components/bottom-navigation/BottomNavigation.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var _useMergedState2 = _interopRequireDefault(require("../../hooks/useMergedState"));

var _class = require("../../utils/class");

var _context = require("./context");

var _utils = require("./utils");

var _BottomNavigationItem = _interopRequireDefault(require("./BottomNavigationItem"));

var _AnimatedBottomNavigationItem = _interopRequireDefault(require("./AnimatedBottomNavigationItem"));

var _excluded = ["key", "style", "className", "itemKey", "linkTo"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var BottomNavigation = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var _classNames;

  var className = props.className,
      style = props.style,
      children = props.children,
      fixed = props.fixed,
      id = props.id,
      onChange = props.onChange,
      activeKey = props.activeKey,
      defaultActiveKey = props.defaultActiveKey,
      zIndex = props.zIndex;
  var prefixCls = (0, _class.getPrefixCls)("bottom-navigation");
  var cls = (0, _clsx.default)(prefixCls, className, (_classNames = {}, _classNames[prefixCls + "-fixed"] = fixed, _classNames));
  var items = (0, _utils.parseItemList)(children);

  if (!items || items.length < 2) {
    console.warn("Bottom Navigation must to have at least 2 items");
  }

  var itemKeys = items.map(function (item) {
    return item.key;
  }).join("_");

  var _useMergedState = (0, _useMergedState2.default)(function () {
    var _items$;

    return (_items$ = items[0]) == null ? void 0 : _items$.key;
  }, {
    value: activeKey,
    defaultValue: defaultActiveKey
  }),
      mergedActiveKey = _useMergedState[0],
      setMergedActiveKey = _useMergedState[1];

  var _useState = (0, _react.useState)(function () {
    return items.findIndex(function (item) {
      return item.key === mergedActiveKey;
    });
  }),
      activeIndex = _useState[0],
      setActiveIndex = _useState[1];

  (0, _react.useEffect)(function () {
    var newActiveIndex = items.findIndex(function (item) {
      return item.key === mergedActiveKey;
    });

    if (newActiveIndex === -1) {
      var _items$newActiveIndex;

      newActiveIndex = Math.max(0, Math.min(activeIndex, items.length - 1));
      setMergedActiveKey((_items$newActiveIndex = items[newActiveIndex]) == null ? void 0 : _items$newActiveIndex.key);
    }

    setActiveIndex(newActiveIndex);
  }, [mergedActiveKey, activeIndex, items, itemKeys, setMergedActiveKey]);

  var handleonChange = function handleonChange(key) {
    var isActiveChanged = key !== mergedActiveKey;
    setMergedActiveKey(key);

    if (isActiveChanged) {
      onChange == null ? void 0 : onChange(key);
    }
  };

  var contextValue = (0, _react.useMemo)(function () {
    return {
      activeKey: mergedActiveKey,
      items: items
    };
  }, [items, mergedActiveKey]);

  var elStyle = _objectSpread(_objectSpread({}, style || {}), zIndex !== undefined ? {
    zIndex: zIndex
  } : {});

  return /*#__PURE__*/_react.default.createElement("div", {
    ref: ref,
    id: id,
    style: elStyle,
    className: cls
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: prefixCls + "-content"
  }, /*#__PURE__*/_react.default.createElement(_context.BottomNavigationContext.Provider, {
    value: contextValue
  }, items.map(function (item) {
    var key = item.key,
        tabStyle = item.style,
        tabClassName = item.className,
        itemKey = item.itemKey,
        linkTo = item.linkTo,
        restTabProps = (0, _objectWithoutPropertiesLoose2.default)(item, _excluded);
    var activeItem = key === mergedActiveKey;

    if (linkTo) {
      return /*#__PURE__*/_react.default.createElement(_AnimatedBottomNavigationItem.default, (0, _extends2.default)({
        key: key
      }, restTabProps, {
        linkTo: linkTo,
        style: tabStyle,
        active: activeItem,
        itemKey: itemKey || key,
        className: tabClassName,
        onChange: handleonChange
      }));
    }

    return /*#__PURE__*/_react.default.createElement(_BottomNavigationItem.default, (0, _extends2.default)({
      key: key
    }, restTabProps, {
      style: tabStyle,
      active: activeItem,
      itemKey: itemKey || key,
      className: tabClassName,
      onChange: handleonChange
    }));
  }))));
});

var _default = BottomNavigation;
exports.default = _default;
