// zmp-ui/cjs/components/radio/Radio.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var _context = _interopRequireDefault(require("./context"));

var _class = require("../../utils/class");

var _excluded = ["className", "children", "style", "disabled", "size", "label", "defaultChecked"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var Radio = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var _classNames, _classNames2;

  var groupContext = (0, _react.useContext)(_context.default);
  var prefixCls = (0, _class.getPrefixCls)("radio");
  var className = props.className,
      children = props.children,
      style = props.style,
      disabled = props.disabled,
      size = props.size,
      label = props.label,
      defaultChecked = props.defaultChecked,
      restProps = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);

  var _useState = (0, _react.useState)("checked" in restProps ? restProps.checked : defaultChecked),
      checked = _useState[0],
      setChecked = _useState[1];

  var radioProps = _objectSpread({}, restProps);

  radioProps.disabled = disabled;

  var onChange = function onChange(e) {
    if (!("checked" in radioProps)) {
      setChecked(e.target.checked);
    }

    props.onChange == null ? void 0 : props.onChange(e);
    groupContext == null ? void 0 : groupContext.onChange == null ? void 0 : groupContext.onChange(e);
  };

  if (groupContext) {
    var _props$value, _groupContext$value;

    radioProps.name = groupContext.name;
    radioProps.checked = ((_props$value = props.value) == null ? void 0 : _props$value.toString()) === ((_groupContext$value = groupContext.value) == null ? void 0 : _groupContext$value.toString());
    radioProps.disabled = radioProps.disabled || groupContext.disabled;
  }

  var radioSize = (groupContext == null ? void 0 : groupContext.size) || size;
  var inputChecked = radioProps.checked !== undefined ? radioProps.checked : checked;
  var wrapperCls = (0, _clsx.default)(prefixCls + "-wrapper", (_classNames = {}, _classNames[prefixCls + "-" + radioSize] = radioSize, _classNames), className);
  var radioCls = (0, _clsx.default)(prefixCls, (_classNames2 = {}, _classNames2[prefixCls + "-checked"] = inputChecked, _classNames2[prefixCls + "-disabled"] = radioProps.disabled, _classNames2));
  return (
    /*#__PURE__*/
    // eslint-disable-next-line jsx-a11y/label-has-associated-control
    _react.default.createElement("label", {
      className: wrapperCls,
      style: style
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: radioCls
    }, /*#__PURE__*/_react.default.createElement("input", (0, _extends2.default)({}, radioProps, {
      checked: !!inputChecked,
      onChange: onChange,
      type: "radio",
      className: prefixCls + "-input",
      ref: ref
    })), /*#__PURE__*/_react.default.createElement("span", {
      className: prefixCls + "-checkmark"
    })), children !== undefined ? /*#__PURE__*/_react.default.createElement("span", null, children) : /*#__PURE__*/_react.default.createElement("span", null, label))
  );
});

var _default = Radio;
exports.default = _default;
