// zmp-ui/esm/components/button/props-type.js
/**
 * @category General
 * @subcategory Button
 * @enum {string}
 */
export var ButtonType;
/**
 * @category General
 * @subcategory Button
 * @enum {string}
 */

(function (ButtonType) {
  ButtonType["highlight"] = "highlight";
  ButtonType["danger"] = "danger";
  ButtonType["neutral"] = "neutral";
})(ButtonType || (ButtonType = {}));

export var ButtonHTMLType;
/**
 * @category General
 * @subcategory Button
 * @enum {string}
 */

(function (ButtonHTMLType) {
  ButtonHTMLType["submit"] = "submit";
  ButtonHTMLType["button"] = "button";
  ButtonHTMLType["reset"] = "reset";
})(ButtonHTMLType || (ButtonHTMLType = {}));

export var ButtonSize;
/**
 * @category General
 * @subcategory Button
 * @enum {string}
 */

(function (ButtonSize) {
  ButtonSize["large"] = "large";
  ButtonSize["medium"] = "medium";
  ButtonSize["small"] = "small";
})(ButtonSize || (ButtonSize = {}));

export var ButtonVariants;

(function (ButtonVariants) {
  ButtonVariants["primary"] = "primary";
  ButtonVariants["secondary"] = "secondary";
  ButtonVariants["tertiary"] = "tertiary";
})(ButtonVariants || (ButtonVariants = {}));
