// zmp-ui/esm/components/radio/Radio.js
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
var _excluded = ["className", "children", "style", "disabled", "size", "label", "defaultChecked"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

import React, { useContext, useState } from "react";
import classNames from "clsx";
import GroupContext from "./context";
import { getPrefixCls } from "../../utils/class";
var Radio = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _classNames, _classNames2;

  var groupContext = useContext(GroupContext);
  var prefixCls = getPrefixCls("radio");

  var className = props.className,
      children = props.children,
      style = props.style,
      disabled = props.disabled,
      size = props.size,
      label = props.label,
      defaultChecked = props.defaultChecked,
      restProps = _objectWithoutPropertiesLoose(props, _excluded);

  var _useState = useState("checked" in restProps ? restProps.checked : defaultChecked),
      checked = _useState[0],
      setChecked = _useState[1];

  var radioProps = _objectSpread({}, restProps);

  radioProps.disabled = disabled;

  var onChange = function onChange(e) {
    if (!("checked" in radioProps)) {
      setChecked(e.target.checked);
    }

    props.onChange == null ? void 0 : props.onChange(e);
    groupContext == null ? void 0 : groupContext.onChange == null ? void 0 : groupContext.onChange(e);
  };

  if (groupContext) {
    var _props$value, _groupContext$value;

    radioProps.name = groupContext.name;
    radioProps.checked = ((_props$value = props.value) == null ? void 0 : _props$value.toString()) === ((_groupContext$value = groupContext.value) == null ? void 0 : _groupContext$value.toString());
    radioProps.disabled = radioProps.disabled || groupContext.disabled;
  }

  var radioSize = (groupContext == null ? void 0 : groupContext.size) || size;
  var inputChecked = radioProps.checked !== undefined ? radioProps.checked : checked;
  var wrapperCls = classNames(prefixCls + "-wrapper", (_classNames = {}, _classNames[prefixCls + "-" + radioSize] = radioSize, _classNames), className);
  var radioCls = classNames(prefixCls, (_classNames2 = {}, _classNames2[prefixCls + "-checked"] = inputChecked, _classNames2[prefixCls + "-disabled"] = radioProps.disabled, _classNames2));
  return (
    /*#__PURE__*/
    // eslint-disable-next-line jsx-a11y/label-has-associated-control
    React.createElement("label", {
      className: wrapperCls,
      style: style
    }, /*#__PURE__*/React.createElement("span", {
      className: radioCls
    }, /*#__PURE__*/React.createElement("input", _extends({}, radioProps, {
      checked: !!inputChecked,
      onChange: onChange,
      type: "radio",
      className: prefixCls + "-input",
      ref: ref
    })), /*#__PURE__*/React.createElement("span", {
      className: prefixCls + "-checkmark"
    })), children !== undefined ? /*#__PURE__*/React.createElement("span", null, children) : /*#__PURE__*/React.createElement("span", null, label))
  );
});
export default Radio;
