// zmp-ui/cjs/components/box/props-type.js
"use strict";

exports.__esModule = true;
exports.VerticalAlignEnum = exports.TextAlignEnum = exports.SkeletonEffect = exports.JustifyContentEnum = exports.FlexDirectionEnum = exports.AlignItemsEnum = exports.AlignContentEnum = void 0;

/* eslint-disable sonarjs/no-duplicate-string */
var SkeletonEffect;
/**
 * @category Layout
 * @subcategory Box
 * @enum {string} VerticalAlign
 */

exports.SkeletonEffect = SkeletonEffect;

(function (SkeletonEffect) {
  SkeletonEffect["fade"] = "fade";
  SkeletonEffect["pulse"] = "pulse";
  SkeletonEffect["wave"] = "wave";
  SkeletonEffect["none"] = "none";
})(SkeletonEffect || (exports.SkeletonEffect = SkeletonEffect = {}));

var VerticalAlignEnum;
exports.VerticalAlignEnum = VerticalAlignEnum;

(function (VerticalAlignEnum) {
  VerticalAlignEnum["baseline"] = "baseline";
  VerticalAlignEnum["sub"] = "sub";
  VerticalAlignEnum["super"] = "super";
  VerticalAlignEnum["top"] = "top";
  VerticalAlignEnum["text-top"] = "text-top";
  VerticalAlignEnum["middle"] = "middle";
  VerticalAlignEnum["bottom"] = "bottom";
  VerticalAlignEnum["text-bottom"] = "text-bottom";
  VerticalAlignEnum["initial"] = "initial";
  VerticalAlignEnum["inherit"] = "inherit";
})(VerticalAlignEnum || (exports.VerticalAlignEnum = VerticalAlignEnum = {}));

var FlexDirectionEnum;
exports.FlexDirectionEnum = FlexDirectionEnum;

(function (FlexDirectionEnum) {
  FlexDirectionEnum["row"] = "row";
  FlexDirectionEnum["row-reverse"] = "row-reverse";
  FlexDirectionEnum["column"] = "column";
  FlexDirectionEnum["column-reverse"] = "column-reverse";
})(FlexDirectionEnum || (exports.FlexDirectionEnum = FlexDirectionEnum = {}));

var JustifyContentEnum;
exports.JustifyContentEnum = JustifyContentEnum;

(function (JustifyContentEnum) {
  JustifyContentEnum["flex-start"] = "flex-start";
  JustifyContentEnum["flex-end"] = "flex-end";
  JustifyContentEnum["center"] = "center";
  JustifyContentEnum["space-between"] = "space-between";
  JustifyContentEnum["space-around"] = "space-around";
  JustifyContentEnum["space-evenly"] = "space-evenly";
  JustifyContentEnum["initial"] = "initial";
})(JustifyContentEnum || (exports.JustifyContentEnum = JustifyContentEnum = {}));

var AlignItemsEnum;
exports.AlignItemsEnum = AlignItemsEnum;

(function (AlignItemsEnum) {
  AlignItemsEnum["flex-start"] = "flex-start";
  AlignItemsEnum["flex-end"] = "flex-end";
  AlignItemsEnum["stretch"] = "stretch";
  AlignItemsEnum["baseline"] = "baseline";
  AlignItemsEnum["center"] = "center";
  AlignItemsEnum["initial"] = "initial";
})(AlignItemsEnum || (exports.AlignItemsEnum = AlignItemsEnum = {}));

var AlignContentEnum;
/**
 * @category Layout
 * @subcategory Box
 * @enum {string} TextAlignEnum
 */

exports.AlignContentEnum = AlignContentEnum;

(function (AlignContentEnum) {
  AlignContentEnum["stretch"] = "stretch";
  AlignContentEnum["center"] = "center";
  AlignContentEnum["flex-start"] = "flex-start";
  AlignContentEnum["flex-end"] = "flex-end";
  AlignContentEnum["space-between"] = "space-between";
  AlignContentEnum["space-around"] = "space-around";
  AlignContentEnum["space-evenly"] = "space-evenly";
  AlignContentEnum["initial"] = "initial";
})(AlignContentEnum || (exports.AlignContentEnum = AlignContentEnum = {}));

var TextAlignEnum;
/**
 * @category Layout
 * @subcategory Box
 * @displayName Box
 *
 */

exports.TextAlignEnum = TextAlignEnum;

(function (TextAlignEnum) {
  TextAlignEnum["left"] = "left";
  TextAlignEnum["right"] = "right";
  TextAlignEnum["center"] = "center";
  TextAlignEnum["justify"] = "justify";
  TextAlignEnum["initial"] = "initital";
  TextAlignEnum["inherit"] = "inherit";
})(TextAlignEnum || (exports.TextAlignEnum = TextAlignEnum = {}));
