// zmp-ui/cjs/components/avatar/avatar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var _propsType = require("./props-type");

var _class = require("../../utils/class");

var _blockIcon = _interopRequireDefault(require("./block-icon"));

var _excluded = ["story", "blocked", "children", "src", "online", "className", "size", "style", "backgroundColor"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var calculateAvatarColor = function calculateAvatarColor(name) {
  if (typeof name !== "string") {
    // eslint-disable-next-line sonarjs/no-duplicate-string
    return _propsType.GradientVariants["GREEN-GREENLIGHT"];
  }

  var sum = 0;
  var length = name.length;

  for (var i = 0; i < length; i += 1) {
    sum += name.charCodeAt(i);
  }

  var colorIndex = sum % Object.keys(_propsType.GradientVariants).length;

  switch (colorIndex) {
    case 1:
      return _propsType.GradientVariants["GREEN-GREENLIGHT"];

    case 2:
      return _propsType.GradientVariants["PURPLE-BLUE"];

    case 3:
      return _propsType.GradientVariants["SKYBLUE-GREEN"];

    default:
      return _propsType.GradientVariants["BLUE-BLUELIGHT"];
  }
};

var Avatar = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var _classNames;

  var story = props.story,
      blocked = props.blocked,
      children = props.children,
      src = props.src,
      online = props.online,
      className = props.className,
      size = props.size,
      style = props.style,
      backgroundColor = props.backgroundColor,
      rest = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  var prefixCls = (0, _class.getPrefixCls)("avatar");
  var bgColor = backgroundColor;

  if (!bgColor && typeof children === "string") {
    bgColor = calculateAvatarColor(children);
  }

  var classes = (0, _clsx.default)(prefixCls, (_classNames = {}, _classNames[prefixCls + "-story"] = story === _propsType.StoryStatus.default, _classNames[prefixCls + "-story-seen"] = story === _propsType.StoryStatus.seen, _classNames[prefixCls + "-text-xsmall"] = size && size <= 24, _classNames[prefixCls + "-text-small"] = size && size > 24 && size <= 32, _classNames[prefixCls + "-text-medium"] = size && size > 32 && size <= 48, _classNames[prefixCls + "-text-large"] = size && size > 48, _classNames[prefixCls + "-color-01"] = bgColor === _propsType.GradientVariants["BLUE-BLUELIGHT"], _classNames[prefixCls + "-color-02"] = bgColor === _propsType.GradientVariants["PURPLE-BLUE"], _classNames[prefixCls + "-color-03"] = bgColor === _propsType.GradientVariants["SKYBLUE-GREEN"], _classNames[prefixCls + "-color-04"] = bgColor === _propsType.GradientVariants["GREEN-GREENLIGHT"], _classNames["" + className] = className, _classNames));
  var avatarStyle = {};

  if (style) {
    avatarStyle = _objectSpread({}, style);
  }

  if (size) {
    avatarStyle = _objectSpread(_objectSpread({}, avatarStyle), {}, {
      "--zaui-avatar-size": size + "px"
    });
  }

  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref,
    className: classes,
    style: avatarStyle
  }, rest), src ? /*#__PURE__*/_react.default.createElement("img", {
    className: prefixCls + "-image",
    src: src,
    alt: ""
  }) : /*#__PURE__*/_react.default.createElement("div", {
    className: prefixCls + "-inner"
  }, children), online && /*#__PURE__*/_react.default.createElement("span", {
    className: prefixCls + "-online"
  }), blocked && /*#__PURE__*/_react.default.createElement("span", {
    className: prefixCls + "-blocked"
  }, /*#__PURE__*/_react.default.createElement(_blockIcon.default, null)));
});

Avatar.displayName = "zaui-avatar";
var _default = Avatar;
exports.default = _default;
