// zmp-ui/cjs/components/snackbar-provider/BaseSnackbar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var _icon = _interopRequireDefault(require("../icon"));

var _class = require("../../utils/class");

var _propsType = require("./props-type");

var _context = _interopRequireDefault(require("./context"));

var _constants = require("./common/constants");

var _progress = _interopRequireDefault(require("../progress"));

var _icons = _interopRequireDefault(require("./icons"));

var _CountDown = _interopRequireDefault(require("./CountDown"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var BaseSnackbar = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var _classNames, _classNames2;

  var text = props.text,
      prefixIcon = props.prefixIcon,
      action = props.action,
      icon = props.icon,
      verticalAction = props.verticalAction,
      style = props.style,
      className = props.className,
      type = props.type,
      position = props.position,
      _props$downloadProgre = props.downloadProgress,
      downloadProgress = _props$downloadProgre === void 0 ? 0 : _props$downloadProgre,
      duration = props.duration,
      zIndex = props.zIndex;
  var snackbarContext = (0, _react.useContext)(_context.default);
  var validPosition = Object.values(_propsType.SnackbarPosition).includes(position) ? position : _constants.DEFAULT_POSITION;
  var prefixCls = (0, _class.getPrefixCls)("snackbar");
  var defaultIcon;

  if (type !== _propsType.SnackbarType.default && type !== _propsType.SnackbarType.countdown) {
    defaultIcon = (0, _icons.default)(type);
  }

  if (type === _propsType.SnackbarType.countdown) {
    defaultIcon = /*#__PURE__*/_react.default.createElement(_CountDown.default, {
      duration: duration || _constants.DEFAULT_COUNTDOWN_DURATION
    });
  }

  var textOnly = !icon && !defaultIcon && !prefixIcon && !action;
  var cls = (0, _clsx.default)(prefixCls, className, (_classNames = {}, _classNames[prefixCls + "-vertical-action"] = verticalAction, _classNames[prefixCls + "-" + validPosition] = validPosition, _classNames[prefixCls + "-text-only"] = textOnly, _classNames));
  var iconEl = prefixIcon || defaultIcon || (icon ? /*#__PURE__*/_react.default.createElement(_icon.default, {
    icon: "zi-info-circle-solid"
  }) : null);
  var progressBar = type === _propsType.SnackbarType.download ? /*#__PURE__*/_react.default.createElement(_progress.default, {
    className: prefixCls + "-progress",
    maxCompleted: 100,
    completed: downloadProgress
  }) : null;

  var onClickAction = function onClickAction(event) {
    event.preventDefault();

    if (action && action.close && snackbarContext) {
      snackbarContext.closeSnackbar == null ? void 0 : snackbarContext.closeSnackbar();
    }

    action == null ? void 0 : action.onClick == null ? void 0 : action.onClick(event);
  };

  var iconClass = (0, _clsx.default)(prefixCls + "-prefix-icon", (_classNames2 = {}, _classNames2[prefixCls + "-success"] = type === _propsType.SnackbarType.success, _classNames2[prefixCls + "-error"] = type === _propsType.SnackbarType.error, _classNames2[prefixCls + "-warning"] = type === _propsType.SnackbarType.warning, _classNames2[prefixCls + "-info"] = type === _propsType.SnackbarType.info, _classNames2[prefixCls + "-download"] = type === _propsType.SnackbarType.download, _classNames2[prefixCls + "-success"] = type === _propsType.SnackbarType.success, _classNames2[prefixCls + "-wifi-connected"] = type === _propsType.SnackbarType["wifi-connected"], _classNames2[prefixCls + "-wifi-disconnected"] = type === _propsType.SnackbarType["wifi-disconnected"], _classNames2[prefixCls + "-loading"] = type === _propsType.SnackbarType.loading, _classNames2));
  var elStyle = {};

  if (style) {
    elStyle = _objectSpread({}, style);
  }

  if (zIndex) {
    elStyle.zIndex = zIndex;
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: cls,
    ref: ref,
    style: elStyle
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: prefixCls + "-content"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: prefixCls + "-desc-wrapper"
  }, iconEl && /*#__PURE__*/_react.default.createElement("div", {
    className: iconClass
  }, iconEl), text && /*#__PURE__*/_react.default.createElement("div", {
    className: prefixCls + "-desc"
  }, text)), progressBar), action && /*#__PURE__*/_react.default.createElement("div", {
    role: "presentation",
    className: prefixCls + "-action",
    onClick: onClickAction
  }, action.text));
});

BaseSnackbar.displayName = "zaui-snackbar";
var _default = BaseSnackbar;
exports.default = _default;
