// zmp-ui/cjs/components/snackbar-provider/props-type.js
"use strict";

exports.__esModule = true;
exports.SnackbarType = exports.SnackbarPosition = void 0;

/* eslint-disable no-use-before-define */

/**
 * @category Feedback
 * @subcategory Snackbar
 * @enum {string}
 */
var SnackbarPosition;
/**
 * @category Feedback
 * @subcategory Snackbar
 * @enum {string}
 */

exports.SnackbarPosition = SnackbarPosition;

(function (SnackbarPosition) {
  SnackbarPosition["top"] = "top";
  SnackbarPosition["bottom"] = "bottom";
})(SnackbarPosition || (exports.SnackbarPosition = SnackbarPosition = {}));

var SnackbarType;
/**
 * @category Feedback
 * @subcategory Snackbar
 * @displayName SnackbarProvider
 */

exports.SnackbarType = SnackbarType;

(function (SnackbarType) {
  SnackbarType["default"] = "default";
  SnackbarType["success"] = "success";
  SnackbarType["info"] = "info";
  SnackbarType["error"] = "error";
  SnackbarType["warning"] = "warning";
  SnackbarType["loading"] = "loading";
  SnackbarType["download"] = "download";
  SnackbarType["countdown"] = "countdown";
  SnackbarType["wifi-connected"] = "wifi-connected";
  SnackbarType["wifi-disconnected"] = "wifi-disconnected";
})(SnackbarType || (exports.SnackbarType = SnackbarType = {})); // NOT INJECT TYPE DECLARE
