// zmp-ui/cjs/components/text/props-type.js
"use strict";

exports.__esModule = true;
exports.TitleTextSize = exports.HeaderTextSize = exports.BodyTextSize = void 0;

/**
 * @category Data Display
 * @subcategory Text
 * @enum {string} BodyTextSize
 */
var BodyTextSize;
/**
 * @category Data Display
 * @subcategory Text
 * @enum {string} TitleTextSize
 */

exports.BodyTextSize = BodyTextSize;

(function (BodyTextSize) {
  BodyTextSize["xLarge"] = "xLarge";
  BodyTextSize["large"] = "large";
  BodyTextSize["normal"] = "normal";
  BodyTextSize["small"] = "small";
  BodyTextSize["xSmall"] = "xSmall";
  BodyTextSize["xxSmall"] = "xxSmall";
  BodyTextSize["xxxSmall"] = "xxxSmall";
  BodyTextSize["xxxxSmall"] = "xxxxSmall";
})(BodyTextSize || (exports.BodyTextSize = BodyTextSize = {}));

var TitleTextSize;
/**
 * @category Data Display
 * @subcategory Text
 * @enum {string} HeaderTextSize
 */

exports.TitleTextSize = TitleTextSize;

(function (TitleTextSize) {
  TitleTextSize["xLarge"] = "xLarge";
  TitleTextSize["large"] = "large";
  TitleTextSize["normal"] = "normal";
  TitleTextSize["small"] = "small";
})(TitleTextSize || (exports.TitleTextSize = TitleTextSize = {}));

var HeaderTextSize;
/**
 * @category Data Display
 * @subcategory Text
 * @displayName Text
 */

exports.HeaderTextSize = HeaderTextSize;

(function (HeaderTextSize) {
  HeaderTextSize["normal"] = "normal";
  HeaderTextSize["small"] = "small";
})(HeaderTextSize || (exports.HeaderTextSize = HeaderTextSize = {})); // NOT INJECT TYPE DECLARE
