// zmp-ui/esm/components/text/props-type.js
/**
 * @category Data Display
 * @subcategory Text
 * @enum {string} BodyTextSize
 */
export var BodyTextSize;
/**
 * @category Data Display
 * @subcategory Text
 * @enum {string} TitleTextSize
 */

(function (BodyTextSize) {
  BodyTextSize["xLarge"] = "xLarge";
  BodyTextSize["large"] = "large";
  BodyTextSize["normal"] = "normal";
  BodyTextSize["small"] = "small";
  BodyTextSize["xSmall"] = "xSmall";
  BodyTextSize["xxSmall"] = "xxSmall";
  BodyTextSize["xxxSmall"] = "xxxSmall";
  BodyTextSize["xxxxSmall"] = "xxxxSmall";
})(BodyTextSize || (BodyTextSize = {}));

export var TitleTextSize;
/**
 * @category Data Display
 * @subcategory Text
 * @enum {string} HeaderTextSize
 */

(function (TitleTextSize) {
  TitleTextSize["xLarge"] = "xLarge";
  TitleTextSize["large"] = "large";
  TitleTextSize["normal"] = "normal";
  TitleTextSize["small"] = "small";
})(TitleTextSize || (TitleTextSize = {}));

export var HeaderTextSize;
/**
 * @category Data Display
 * @subcategory Text
 * @displayName Text
 */

(function (HeaderTextSize) {
  HeaderTextSize["normal"] = "normal";
  HeaderTextSize["small"] = "small";
})(HeaderTextSize || (HeaderTextSize = {})); // NOT INJECT TYPE DECLARE
