// zmp-ui/cjs/components/sheet/action-sheet.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var _button = _interopRequireDefault(require("../button"));

var _class = require("../../utils/class");

var _sheet = _interopRequireDefault(require("./sheet"));

var _excluded = ["className", "actions", "title"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var Action = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var className = props.className,
      actions = props.actions,
      title = props.title,
      rest = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  var sheetRef = (0, _react.useRef)(null);
  var prefixCls = (0, _class.getPrefixCls)("action-sheet");
  var groups = actions || [];

  if (!Array.isArray(groups[0])) {
    groups = [groups.filter(function (group) {
      return !Array.isArray(group);
    })];
  }

  (0, _react.useImperativeHandle)(ref, function () {
    var _sheetRef$current;

    return (_sheetRef$current = sheetRef.current) == null ? void 0 : _sheetRef$current.sheet;
  });
  var cls = (0, _clsx.default)(prefixCls, {}, className);
  var actionsChildren = groups.map(function (group, indx) {
    return /*#__PURE__*/_react.default.createElement("ul", {
      className: prefixCls + "-actions-group" // eslint-disable-next-line react/no-array-index-key
      ,
      key: "zaui-modal-action-group-key-" + indx
    }, group.map(function (action, index) {
      var _classNames;

      var text = action.text,
          onClick = action.onClick,
          actionClassName = action.className,
          style = action.style,
          close = action.close,
          highLight = action.highLight,
          danger = action.danger,
          disabled = action.disabled;
      var onClose = props.onClose;
      var btnCls = (0, _clsx.default)(actionClassName, prefixCls + "-action", (_classNames = {}, _classNames[prefixCls + "-action-highlight"] = highLight, _classNames));

      var onActionClick = function onActionClick(e) {
        if (close && onClose) {
          onClose(e);
          return;
        }

        if (onClick) {
          onClick(e);
        }
      };

      var actionType = "neutral";

      if (highLight) {
        actionType = "highlight";
      }

      if (danger) {
        actionType = "danger";
      }

      return /*#__PURE__*/_react.default.createElement("li", {
        className: btnCls,
        key: action.key || "zaui-modal-action-key-" + index
      }, /*#__PURE__*/_react.default.createElement(_button.default, {
        type: actionType,
        style: style,
        variant: "tertiary",
        disabled: disabled,
        fullWidth: true,
        className: prefixCls + "-action-button",
        onClick: onActionClick
      }, text));
    }));
  }) || null;
  return /*#__PURE__*/_react.default.createElement(_sheet.default, (0, _extends2.default)({
    className: cls,
    ref: sheetRef,
    handler: false,
    title: undefined
  }, rest), title && /*#__PURE__*/_react.default.createElement("div", {
    className: prefixCls + "-title-wrapper"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: prefixCls + "-title"
  }, title)), actionsChildren && /*#__PURE__*/_react.default.createElement("div", {
    className: prefixCls + "-actions-groups"
  }, actionsChildren));
});

Action.displayName = "zaui-action-sheet";
var _default = Action;
exports.default = _default;
