// zmp-ui/cjs/components/snackbar-provider/common/constants.js
"use strict";

exports.__esModule = true;
exports.DEFAULT_POSITION = exports.DEFAULT_LOADING_DURATION = exports.DEFAULT_INTERVAL = exports.DEFAULT_DURATION = exports.DEFAULT_DOWNLOADING_DURATION = exports.DEFAULT_COUNTDOWN_DURATION = exports.COUNTDOWN_STROKE_WIDTH = exports.COUNTDOWN_SIZE = exports.COUNTDOWN_RADIUS = void 0;

var _propsType = require("../props-type");

var DEFAULT_DURATION = 2000;
exports.DEFAULT_DURATION = DEFAULT_DURATION;
var DEFAULT_POSITION = _propsType.SnackbarPosition.bottom;
exports.DEFAULT_POSITION = DEFAULT_POSITION;
var DEFAULT_INTERVAL = 250;
exports.DEFAULT_INTERVAL = DEFAULT_INTERVAL;
var DEFAULT_COUNTDOWN_DURATION = 5000;
exports.DEFAULT_COUNTDOWN_DURATION = DEFAULT_COUNTDOWN_DURATION;
var DEFAULT_LOADING_DURATION = 5000;
exports.DEFAULT_LOADING_DURATION = DEFAULT_LOADING_DURATION;
var DEFAULT_DOWNLOADING_DURATION = 5000;
exports.DEFAULT_DOWNLOADING_DURATION = DEFAULT_DOWNLOADING_DURATION;
var COUNTDOWN_SIZE = 24;
exports.COUNTDOWN_SIZE = COUNTDOWN_SIZE;
var COUNTDOWN_STROKE_WIDTH = 2;
exports.COUNTDOWN_STROKE_WIDTH = COUNTDOWN_STROKE_WIDTH;
var COUNTDOWN_RADIUS = 12;
exports.COUNTDOWN_RADIUS = COUNTDOWN_RADIUS;
