// zmp-ui/esm/components/box/props-type.js
/* eslint-disable sonarjs/no-duplicate-string */
export var SkeletonEffect;
/**
 * @category Layout
 * @subcategory Box
 * @enum {string} VerticalAlign
 */

(function (SkeletonEffect) {
  SkeletonEffect["fade"] = "fade";
  SkeletonEffect["pulse"] = "pulse";
  SkeletonEffect["wave"] = "wave";
  SkeletonEffect["none"] = "none";
})(SkeletonEffect || (SkeletonEffect = {}));

export var VerticalAlignEnum;

(function (VerticalAlignEnum) {
  VerticalAlignEnum["baseline"] = "baseline";
  VerticalAlignEnum["sub"] = "sub";
  VerticalAlignEnum["super"] = "super";
  VerticalAlignEnum["top"] = "top";
  VerticalAlignEnum["text-top"] = "text-top";
  VerticalAlignEnum["middle"] = "middle";
  VerticalAlignEnum["bottom"] = "bottom";
  VerticalAlignEnum["text-bottom"] = "text-bottom";
  VerticalAlignEnum["initial"] = "initial";
  VerticalAlignEnum["inherit"] = "inherit";
})(VerticalAlignEnum || (VerticalAlignEnum = {}));

export var FlexDirectionEnum;

(function (FlexDirectionEnum) {
  FlexDirectionEnum["row"] = "row";
  FlexDirectionEnum["row-reverse"] = "row-reverse";
  FlexDirectionEnum["column"] = "column";
  FlexDirectionEnum["column-reverse"] = "column-reverse";
})(FlexDirectionEnum || (FlexDirectionEnum = {}));

export var JustifyContentEnum;

(function (JustifyContentEnum) {
  JustifyContentEnum["flex-start"] = "flex-start";
  JustifyContentEnum["flex-end"] = "flex-end";
  JustifyContentEnum["center"] = "center";
  JustifyContentEnum["space-between"] = "space-between";
  JustifyContentEnum["space-around"] = "space-around";
  JustifyContentEnum["space-evenly"] = "space-evenly";
  JustifyContentEnum["initial"] = "initial";
})(JustifyContentEnum || (JustifyContentEnum = {}));

export var AlignItemsEnum;

(function (AlignItemsEnum) {
  AlignItemsEnum["flex-start"] = "flex-start";
  AlignItemsEnum["flex-end"] = "flex-end";
  AlignItemsEnum["stretch"] = "stretch";
  AlignItemsEnum["baseline"] = "baseline";
  AlignItemsEnum["center"] = "center";
  AlignItemsEnum["initial"] = "initial";
})(AlignItemsEnum || (AlignItemsEnum = {}));

export var AlignContentEnum;
/**
 * @category Layout
 * @subcategory Box
 * @enum {string} TextAlignEnum
 */

(function (AlignContentEnum) {
  AlignContentEnum["stretch"] = "stretch";
  AlignContentEnum["center"] = "center";
  AlignContentEnum["flex-start"] = "flex-start";
  AlignContentEnum["flex-end"] = "flex-end";
  AlignContentEnum["space-between"] = "space-between";
  AlignContentEnum["space-around"] = "space-around";
  AlignContentEnum["space-evenly"] = "space-evenly";
  AlignContentEnum["initial"] = "initial";
})(AlignContentEnum || (AlignContentEnum = {}));

export var TextAlignEnum;
/**
 * @category Layout
 * @subcategory Box
 * @displayName Box
 *
 */

(function (TextAlignEnum) {
  TextAlignEnum["left"] = "left";
  TextAlignEnum["right"] = "right";
  TextAlignEnum["center"] = "center";
  TextAlignEnum["justify"] = "justify";
  TextAlignEnum["initial"] = "initital";
  TextAlignEnum["inherit"] = "inherit";
})(TextAlignEnum || (TextAlignEnum = {}));
